build/lib/src/library_cycle_graph/library_cycle_graph_loader.dart

```diff
@@ -0,0 +1,395 @@
+  /// expire after [upToPhase]--are added to [_graphsToComputeByPhase] to
+  /// be completed later.
+  /// [_graphsToComputeByPhase].
+  void _buildCycles(int upToPhase) {
+    // Process phases that have work to do in ascending order.
+    while (true) {
+      int phase;
+      Set<AssetId> idsToComputeCyclesFrom;
+      if (_newAssets.isNotEmpty) {
+        // New assets: work to do at phase 0, the initial build state.
+        phase = 0;
+        idsToComputeCyclesFrom = _newAssets;
+        _newAssets = {};
+      } else {
+        // Work through phases <= `upToPhase` at which graphs expire,
+        // so there are new values to compute.
+        if (_graphsToComputeByPhase.isEmpty) break;
+        phase = _graphsToComputeByPhase.keys.reduce(min);
+        if (phase > upToPhase) break;
+        idsToComputeCyclesFrom = _graphsToComputeByPhase.remove(phase)!;
+      }
+
+      // Edges for strongly connected components computation.
+      Iterable<AssetId> edgesFromId(AssetId id) {
+        final deps = _assetDeps[id]!.valueAt(phase: phase).deps;
+
+        // Check edge against cycles that have already been computed
+        // at the current `phase`. Newly discovered assets at the same phase
+        // cannot be part of already-computed cycles, so prevent recomputation
+        // of those same cycles by hiding deps onto them.
+        return deps.where(
+          (id) => _cycles[id]?.isExpiredAt(phase: phase) ?? true,
+        );
+      }
+
+      // Do the strongly connected components computation and convert
+      // from its output, a list of lists of IDs, to a list of [LibraryCycle].
+      final newComponentLists = stronglyConnectedComponents(
+        idsToComputeCyclesFrom,
+        edgesFromId,
+      );
+      final newCycles =
+          newComponentLists.map((list) {
+            // Compare to the library cycle computed at `phase - 1`. If the
+            // cycles are the same size then they must have the same contents,
+            // because cycles only change by growing as phases progress. In that
+            // case, reuse the existing [LibraryCycle].
+            final maybePhasedCycle = _cycles[list.first];
+            if (maybePhasedCycle != null) {
+              final value = maybePhasedCycle.valueAt(phase: phase - 1);
+              if (value.ids.length == list.length) {
+                return value;
+              }
+            }
+            // The cycle is new or has changed, return a new value.
+            return LibraryCycle((b) => b..ids.replace(list));
+          }).toList();
+
+      // Build graphs from cycles.
+      _buildGraphs(phase, newCycles: newCycles);
+
+      for (final cycle in newCycles) {
+        // A cycle expires at the earliest expiry phase of all its transitive
+        // deps, because new deps of any transitive dep might change the
+        // cycle. Get this from the graph built by `_buildGraphs`.
+        final expiresAt =
```

Comment:
There are still some `expiresAt` (and below some `expiresAfter: expiresAt`) - maybe they should be renamed too.